raft/fsm.cc

```diff
@@ -1019,8 +1019,12 @@ bool fsm::apply_snapshot(snapshot_descriptor snp, size_t max_trailing_entries, s
     // If the snapshot is local, _commit_idx is larger than snp.idx.
     // Otherwise snp.idx becomes the new commit index.
     _commit_idx = std::max(_commit_idx, snp.idx);
-    _output.snp.emplace(fsm_output::applied_snapshot{snp, local, max_trailing_entries});
-    size_t units = _log.apply_snapshot(std::move(snp), max_trailing_entries, max_trailing_bytes);
+    auto [units, new_first_index] =
+        _log.apply_snapshot(std::move(snp), max_trailing_entries, max_trailing_bytes);
+    _output.snp.emplace(fsm_output::applied_snapshot{
+        .snp = _log.get_snapshot(),
+        .is_local = local,
+        .preserved_log_entries = (_log.get_snapshot().idx + 1) - new_first_index});
```

Comment:
Why the +1?

Suppose we want to truncate everything behind the snapshot (preserve 0 entries).

So we want new_first_index = snapshot.idx, and preserved_log_entries = 0.

But this would give preserved_log_entries = 1 in such case.

Reply:
That's not very obvious, but if we preserved 0 entries the first index in the log is `snap.idx + 1`, see the comment at raft/log.hh:30

Reply:
So the entry with the index equal to snapshot index is also considered a "trailing entry"

Reply:
I think we should specify that somewhere in a comment, right now it's not obviously stated anywhere it seems
Maybe at `apply_snapshot` declaration in raft/log.hh

it should say what a trailing entry is: it is any entry that remains with index smaller **or equal** to the snapshot index

Reply:
And also in the comment specify that if there are no trailing entries, then the first index will be snapshot index + 1 (even if there is actually no entry at that index, in case the log now contains only the snapshot)

Reply:
Ok, will add a comment.

Reply:
Added a comment in the new version